// okx-api/dist/cjs/constants/funding.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ASSET_BILL_TYPE = void 0;
/**
 * Asset bill type, auto-generated using table from funding asset bill details:
 * https://www.okx.com/docs-v5/en/#rest-api-funding-asset-bills-details
 */
var ASSET_BILL_TYPE;
(function (ASSET_BILL_TYPE) {
    ASSET_BILL_TYPE["DEPOSIT"] = "1";
    ASSET_BILL_TYPE["WITHDRAWAL"] = "2";
    ASSET_BILL_TYPE["CANCELED_WITHDRAWAL"] = "13";
    ASSET_BILL_TYPE["TRANSFER_TO_SUB_ACCOUNT"] = "20";
    ASSET_BILL_TYPE["TRANSFER_FROM_SUB_ACCOUNT"] = "21";
    ASSET_BILL_TYPE["CLAIM"] = "28";
    ASSET_BILL_TYPE["SYSTEM_REVERSAL"] = "47";
    ASSET_BILL_TYPE["RECEIVED_FROM_ACTIVITIES"] = "48";
    ASSET_BILL_TYPE["GIVEN_AWAY_TO_ACTIVITIES"] = "49";
    ASSET_BILL_TYPE["RECEIVED_FROM_APPOINTMENTS"] = "50";
    ASSET_BILL_TYPE["DEDUCTED_FROM_APPOINTMENTS"] = "51";
    ASSET_BILL_TYPE["RED_PACKET_SENT"] = "52";
    ASSET_BILL_TYPE["RED_PACKET_SNATCHED"] = "53";
    ASSET_BILL_TYPE["RED_PACKET_REFUNDED"] = "54";
    ASSET_BILL_TYPE["CONVERSION"] = "61";
    ASSET_BILL_TYPE["CLAIM_REBATE_CARD"] = "68";
    ASSET_BILL_TYPE["DISTRIBUTE_REBATE_CARD"] = "69";
    ASSET_BILL_TYPE["TOKEN_RECEIVED"] = "72";
    ASSET_BILL_TYPE["TOKEN_GIVEN_AWAY"] = "73";
    ASSET_BILL_TYPE["TOKEN_REFUNDED"] = "74";
    ASSET_BILL_TYPE["SUBSCRIPTION_TO_SAVINGS"] = "75";
    ASSET_BILL_TYPE["REDEMPTION_TO_SAVINGS"] = "76";
    ASSET_BILL_TYPE["DISTRIBUTE"] = "77";
    ASSET_BILL_TYPE["LOCK_UP"] = "78";
    ASSET_BILL_TYPE["NODE_VOTING"] = "79";
    ASSET_BILL_TYPE["STAKING_1"] = "80";
    ASSET_BILL_TYPE["VOTE_REDEMPTION"] = "81";
    ASSET_BILL_TYPE["STAKING_REDEMPTION_1"] = "82";
    ASSET_BILL_TYPE["STAKING_YIELD"] = "83";
    ASSET_BILL_TYPE["VIOLATION_FEE"] = "84";
    ASSET_BILL_TYPE["POW_MINING_YIELD"] = "85";
    ASSET_BILL_TYPE["CLOUD_MINING_PAY"] = "86";
    ASSET_BILL_TYPE["CLOUD_MINING_YIELD"] = "87";
    ASSET_BILL_TYPE["SUBSIDY"] = "88";
    ASSET_BILL_TYPE["STAKING_2"] = "89";
    ASSET_BILL_TYPE["STAKING_SUBSCRIPTION"] = "90";
    ASSET_BILL_TYPE["STAKING_REDEMPTION_2"] = "91";
    ASSET_BILL_TYPE["ADD_COLLATERAL"] = "92";
    ASSET_BILL_TYPE["REDEEM_COLLATERAL"] = "93";
    ASSET_BILL_TYPE["INVESTMENT"] = "94";
    ASSET_BILL_TYPE["BORROWER_BORROWS"] = "95";
    ASSET_BILL_TYPE["PRINCIPAL_TRANSFERRED_IN"] = "96";
    ASSET_BILL_TYPE["BORROWER_TRANSFERRED_LOAN_OUT"] = "97";
    ASSET_BILL_TYPE["BORROWER_TRANSFERRED_INTEREST_OUT"] = "98";
    ASSET_BILL_TYPE["INVESTOR_TRANSFERRED_INTEREST_IN"] = "99";
    ASSET_BILL_TYPE["PREPAYMENT_PENALTY_TRANSFERRED_IN"] = "102";
    ASSET_BILL_TYPE["PREPAYMENT_PENALTY_TRANSFERRED_OUT"] = "103";
    ASSET_BILL_TYPE["FEE_TRANSFERRED_IN"] = "104";
    ASSET_BILL_TYPE["FEE_TRANSFERRED_OUT"] = "105";
    ASSET_BILL_TYPE["OVERDUE_FEE_TRANSFERRED_IN"] = "106";
    ASSET_BILL_TYPE["OVERDUE_FEE_TRANSFERRED_OUT"] = "107";
    ASSET_BILL_TYPE["OVERDUE_INTEREST_TRANSFERRED_OUT"] = "108";
    ASSET_BILL_TYPE["OVERDUE_INTEREST_TRANSFERRED_IN"] = "109";
    ASSET_BILL_TYPE["COLLATERAL_FOR_CLOSED_POSITION_TRANSFERRED_IN"] = "110";
    ASSET_BILL_TYPE["COLLATERAL_FOR_CLOSED_POSITION_TRANSFERRED_OUT"] = "111";
    ASSET_BILL_TYPE["COLLATERAL_FOR_LIQUIDATION_TRANSFERRED_IN"] = "112";
    ASSET_BILL_TYPE["COLLATERAL_FOR_LIQUIDATION_TRANSFERRED_OUT"] = "113";
    ASSET_BILL_TYPE["INSURANCE_FUND_TRANSFERRED_IN"] = "114";
    ASSET_BILL_TYPE["INSURANCE_FUND_TRANSFERRED_OUT_"] = "115";
    ASSET_BILL_TYPE["PLACE_AN_ORDER"] = "116";
    ASSET_BILL_TYPE["FULFILL_AN_ORDER"] = "117";
    ASSET_BILL_TYPE["CANCEL_AN_ORDER"] = "118";
    ASSET_BILL_TYPE["MERCHANTS_UNLOCK_DEPOSIT"] = "119";
    ASSET_BILL_TYPE["MERCHANTS_ADD_DEPOSIT"] = "120";
    ASSET_BILL_TYPE["FIATGATEWAY_PLACE_AN_ORDER"] = "121";
    ASSET_BILL_TYPE["FIATGATEWAY_CANCEL_AN_ORDER"] = "122";
    ASSET_BILL_TYPE["FIATGATEWAY_FULFILL_AN_ORDER"] = "123";
    ASSET_BILL_TYPE["JUMPSTART_UNLOCKING"] = "124";
    ASSET_BILL_TYPE["MANUAL_DEPOSIT"] = "125";
    ASSET_BILL_TYPE["INTEREST_DEPOSIT"] = "126";
    ASSET_BILL_TYPE["INVESTMENT_FEE_TRANSFERRED_IN"] = "127";
    ASSET_BILL_TYPE["INVESTMENT_FEE_TRANSFERRED_OUT"] = "128";
    ASSET_BILL_TYPE["REWARDS_TRANSFERRED_IN"] = "129";
    ASSET_BILL_TYPE["TRANSFERRED_FROM_TRADING_ACCOUNT"] = "130";
    ASSET_BILL_TYPE["TRANSFERRED_TO_TRADING_ACCOUNT"] = "131";
    ASSET_BILL_TYPE["FROZEN_BY_CUSTOMER_SERVICE"] = "132";
    ASSET_BILL_TYPE["UNFROZEN_BY_CUSTOMER_SERVICE"] = "133";
    ASSET_BILL_TYPE["TRANSFERRED_BY_CUSTOMER_SERVICE"] = "134";
    ASSET_BILL_TYPE["CROSS_CHAIN_EXCHANGE"] = "135";
    ASSET_BILL_TYPE["CONVERT"] = "136";
    ASSET_BILL_TYPE["ETH_2_0_SUBSCRIPTION"] = "137";
    ASSET_BILL_TYPE["ETH_2_0_SWAPPING"] = "138";
    ASSET_BILL_TYPE["ETH_2_0_EARNINGS"] = "139";
    ASSET_BILL_TYPE["FEE_TRANSFER_IN"] = "141";
    ASSET_BILL_TYPE["FEE_TRANSFER_OUT"] = "142";
    ASSET_BILL_TYPE["SYSTEM_REVERSE"] = "143";
    ASSET_BILL_TYPE["TRANSFER_OUT_OF_UNIFIED_ACCOUNT_RESERVE"] = "144";
    ASSET_BILL_TYPE["REWARD_EXPIRED"] = "145";
    ASSET_BILL_TYPE["CUSTOMER_FEEDBACK"] = "146";
    ASSET_BILL_TYPE["VESTED_SUSHI_REWARDS"] = "147";
    ASSET_BILL_TYPE["AFFILIATE_COMMISSION"] = "150";
    ASSET_BILL_TYPE["REFERRAL_REWARD"] = "151";
    ASSET_BILL_TYPE["BROKER_REWARD"] = "152";
    ASSET_BILL_TYPE["JOINER_REWARD"] = "153";
    ASSET_BILL_TYPE["MYSTERY_BOX_REWARD"] = "154";
    ASSET_BILL_TYPE["REWARDS_WITHDRAW"] = "155";
    ASSET_BILL_TYPE["FEE_REBATE_1"] = "156";
    ASSET_BILL_TYPE["COLLECTED_CRYPTO"] = "157";
    ASSET_BILL_TYPE["DUAL_INVESTMENT_SUBSCRIBE"] = "160";
    ASSET_BILL_TYPE["DUAL_INVESTMENT_COLLECTION"] = "161";
    ASSET_BILL_TYPE["DUAL_INVESTMENT_PROFIT"] = "162";
    ASSET_BILL_TYPE["DUAL_INVESTMENT_REFUND"] = "163";
    ASSET_BILL_TYPE["NEW_YEAR_REWARDS_2022"] = "169";
    ASSET_BILL_TYPE["REFERRAL_REFERRED_CUSTOMER"] = "170";
    ASSET_BILL_TYPE["REFERRAL_WAS_REFERRED"] = "171";
    ASSET_BILL_TYPE["SUB_AFFILIATE_COMMISSION"] = "172";
    ASSET_BILL_TYPE["FEE_REBATE_2"] = "173";
    ASSET_BILL_TYPE["PAY"] = "174";
    ASSET_BILL_TYPE["LOCKED_COLLATERAL"] = "175";
    ASSET_BILL_TYPE["LOAN"] = "176";
    ASSET_BILL_TYPE["ADDED_COLLATERAL"] = "177";
    ASSET_BILL_TYPE["RETURNED_COLLATERAL"] = "178";
    ASSET_BILL_TYPE["REPAYMENT"] = "179";
    ASSET_BILL_TYPE["UNLOCKED_COLLATERAL"] = "180";
    ASSET_BILL_TYPE["REPAY_AIRDROPPED_CRYPTO"] = "181";
    ASSET_BILL_TYPE["FEEDBACK_BOUNTY"] = "182";
    ASSET_BILL_TYPE["INVITE_FRIENDS_REWARDS"] = "183";
    ASSET_BILL_TYPE["DIVIDE_THE_REWARD_POOL"] = "184";
    ASSET_BILL_TYPE["BROKER_CONVERT_REWARD"] = "185";
    ASSET_BILL_TYPE["FREEETH"] = "186";
    ASSET_BILL_TYPE["CONVERT_TRANSFER"] = "187";
    ASSET_BILL_TYPE["SLOT_AUCTION_CONVERSION"] = "188";
    ASSET_BILL_TYPE["MYSTERY_BOX_BONUS"] = "189";
    ASSET_BILL_TYPE["CARD_PAYMENT_BUY_1"] = "193";
    ASSET_BILL_TYPE["CARD_PAYMENT_BUY_2"] = "194";
    ASSET_BILL_TYPE["UNTRADABLE_ASSET_WITHDRAWAL"] = "195";
    ASSET_BILL_TYPE["UNTRADABLE_ASSET_WITHDRAWAL_REVOKED"] = "196";
    ASSET_BILL_TYPE["UNTRADABLE_ASSET_DEPOSIT"] = "197";
    ASSET_BILL_TYPE["UNTRADABLE_ASSET_COLLECTION_1"] = "198";
    ASSET_BILL_TYPE["UNTRADABLE_ASSET_COLLECTION_2"] = "199";
    ASSET_BILL_TYPE["BUY"] = "200";
    ASSET_BILL_TYPE["WIN_CRYPTO_WITH_SATOSHI"] = "211";
})(ASSET_BILL_TYPE || (exports.ASSET_BILL_TYPE = ASSET_BILL_TYPE = {}));
